File: deck/seed/code/native/node/process/runtime/prompt.ts
// Line reading for node, over node:readline. The interface is held rather than rebuilt per call: a readline
// interface takes the terminal, and two of them over one stdin fight for the keystrokes.
//
// Reached only through the public process/line API.
import * as readline from 'node:readline'

const prompt = {
  lineOpen: async (): Promise<readline.Interface> =>
    readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    }),

  lineAsk: async (
    tool: readline.Interface,
    text: string,
  ): Promise<string> =>
    new Promise(answer => {
      tool.question(text, line => answer(line))
    }),

  // the next line, or '' at the end of input
  lineRead: async (tool: readline.Interface): Promise<string> =>
    new Promise(answer => {
      const done = (line?: string): void => {
        tool.off('line', onLine)
        tool.off('close', onClose)
        answer(line ?? '')
      }
      const onLine = (line: string): void => done(line)
      const onClose = (): void => done()

      tool.once('line', onLine)
      tool.once('close', onClose)
    }),

  lineClose: async (tool: readline.Interface): Promise<void> => {
    tool.close()
  },
}

File: deck/seed/code/native/swift/process/runtime/prompt.swift
// Line reading for the swift target, over Foundation's standard input. Reached only through the public
// process/line API.
//
// `readLine` is synchronous and there is no asynchronous stdin in Foundation, so the read runs on a detached task
// and the caller's task is not blocked. That is the honest shape: waiting for a person to type is not something
// the kernel can hand back sooner.
import Foundation

enum prompt {
  // the reader holds nothing on this backend (readLine owns the buffer), but the handle exists so the API is the
  // same one every other target has
  final class Tool: @unchecked Sendable {
    var open = true
  }

  static func lineOpen() async -> Tool {
    Tool()
  }

  static func lineAsk(_ tool: Tool, _ prompt: String) async -> String {
    FileHandle.standardOutput.write(Data(prompt.utf8))

    return await lineRead(tool)
  }

  // the line without its newline, or "" at the end of input
  static func lineRead(_ tool: Tool) async -> String {
    guard tool.open else { return "" }

    return await Task.detached { readLine(strippingNewline: true) ?? "" }.value
  }

  static func lineClose(_ tool: Tool) async {
    tool.open = false
  }
}
